### ISSUE TITLE
[BUG] prefill fails on PostgreSQL 9.6.2

### ISSUE THREAD

**Issue by mikemcgowan** [bolt/bolt#6649]
Attempting to prefill an empty database with the Lorem Ipsum test data fails on PostgreSQL 9.6.2

 - Relevant Bolt Version: 3.2
 - Install type: GitHub checkout
 - PHP version: 7.1
 - Used web server: Nginx

Requesting http://localhost/bolt/prefill?force=1 results in the following exception:

```
An exception occurred while executing 'INSERT INTO bolt_entries (datechanged, title, slug, teaser, body, image, video, status, templatefields, datecreated, datepublish, datedepublish, ownerid) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)' with params ["2017-05-10 16:31:48", "Ut optime, secundum naturam affectum esse possit.", "ut-optime-secundum-naturam-affectum-esse-possit", "<p>Lorem ipsum dolor sit amet, consectetur adipiscing elit. <mark>Explanetur igitur.<\/mark> Terram, mihi crede, ea lanx et maria deprimet. <\/p>", "<p>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Non semper, inquam; Negat enim summo bono afferre incrementum diem. <a href='http:\/\/loripsum.net\/' target='_blank'>Dici enim nihil potest verius.<\/a> <b>Beatus sibi videtur esse moriens.<\/b> Primum in nostrane potestate est, quid meminerimus? Ostendit pedes et pectus. <\/p>\n\n<p>Istam voluptatem perpetuam quis potest praestare sapienti? Minime vero istorum quidem, inquit. Ait enim se, si uratur, Quam hoc suave! dicturum. De vacuitate doloris eadem sententia erit. Collige omnia, quae soletis: Praesidium amicorum. Maximas vero virtutes iacere omnis necesse est voluptate dominante. Huius, Lyco, oratione locuples, rebus ipsis ielunior. Si mala non sunt, iacet omnis ratio Peripateticorum. <\/p>\n\n<ol>\n\t<li>Intellegi quidem, ut propter aliam quampiam rem, verbi gratia propter voluptatem, nos amemus;<\/li>\n\t<li>Et quidem illud ipsum non nimium probo et tantum patior, philosophum loqui de cupiditatibus finiendis.<\/li>\n\t<li>Unum nescio, quo modo possit, si luxuriosus sit, finitas cupiditates habere.<\/li>\n\t<li>Quae in controversiam veniunt, de iis, si placet, disseramus.<\/li>\n\t<li>Ratio quidem vestra sic cogit.<\/li>\n\t<li>Quod autem ratione actum est, id officium appellamus.<\/li>\n<\/ol>\n\n\n<p>Duo Reges: constructio interrete. Et quod est munus, quod opus sapientiae? <b>Bonum valitudo: miser morbus.<\/b> Uterque enim summo bono fruitur, id est voluptate. Summus dolor plures dies manere non potest? Satisne vobis videor pro meo iure in vestris auribus commentatus? Non laboro, inquit, de nomine. Quamquam tu hanc copiosiorem etiam soles dicere. Sed haec omittamus; Quae ista amicitia est? <\/p>\n\n<ul>\n\t<li>Egone quaeris, inquit, quid sentiam?<\/li>\n\t<li>Aliter homines, aliter philosophos loqui putas oportere?<\/li>\n\t<li>Pudebit te, inquam, illius tabulae, quam Cleanthes sane commode verbis depingere solebat.<\/li>\n<\/ul>", "{\"file\":\"bell-bills-cash-register-2738.jpg\"}", "", "published", "[]", "2017-05-10 16:31:48", "2016-06-26 18:54:02", null, 1]:

SQLSTATE[22P02]: Invalid text representation: 7 ERROR: invalid input syntax for type json
DETAIL: The input string ended unexpectedly.
CONTEXT: JSON data, line 1:
```

Basically, it's trying to insert empty string as the JSON value for the `video` field. PostgreSQL 9.6.2 apparently doesn't like that. If I try to do it manually in `psql`, I get:

```
bolt=# insert into bolt_entries (slug, datecreated, datechanged, status, image) values ('foo', '2017-05-10 16:31:48', '2017-05-10 16:31:48', 'published', '');
ERROR:  invalid input syntax for type json
LINE 1: ...17-05-10 16:31:48', '2017-05-10 16:31:48', 'published', '');
                                                                   ^
DETAIL:  The input string ended unexpectedly.
CONTEXT:  JSON data, line 1: 
```

In other words, the same error.

**Comment by GawainLynch** [comment]
Which 3.2.x? I fixed some of this in #6548 … just wondering if this is more.

Funnily enough I am working on some PostgreSQL stuff right this moment :laughing:

**Comment by mikemcgowan** [comment]
3.2.13

HEAD of my checked out branch is 14a518f94f0ee705ba5024c8f79c498e3b6dba3c

```
commit 14a518f94f0ee705ba5024c8f79c498e3b6dba3c
Merge: 6a0bd01 9c040a8
Author: Bob den Otter <bobdenotter@gmail.com>
Date:   Wed May 10 13:09:13 2017 +0200

    Merge pull request #6644 from GawainLynch/hotfix/embed
    
    [3.2] Embed updates

 src/Embed/GuzzleDispatcher.php            | 120 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++--------------
 src/Embed/Resolver.php                    |  45 ++++++++++++++++++++++++++++++---------------
 src/Provider/EmbedServiceProvider.php     |  23 +++++++++++++++++++----
 tests/phpunit/unit/Embed/ResolverTest.php |   2 +-
 4 files changed, 156 insertions(+), 34 deletions(-)
```

**Comment by mikemcgowan** [comment]
Ahh, sorry, I guess I didn't search _closed_ issues ...

**Comment by GawainLynch** [comment]
> Ahh, sorry, I guess I didn't search closed issues ...

Zero stress, just checking. That is the `HEAD` of `release/3.2` now, so there is another problem … I'll try to dig deeper today :+1:

**Comment by mikemcgowan** [comment]
It seems to me that the default value for any field that gets created as type `json` in PostgreSQL should be `{}` and not empty string. The [content type fields that map to JSON](https://github.com/bolt/bolt/blob/release/3.2/src/Storage/Database/Schema/Table/ContentType.php#L33) are:

```
        // JSON arrays
        'filelist'       => 'columnJson',
        'geolocation'    => 'columnJson',
        'image'          => 'columnJson',
        'imagelist'      => 'columnJson',
        'selectmultiple' => 'columnJson',
        'templatefields' => 'columnJson',
        'video'          => 'columnJson',
```

The below patch fixes this bug. I could open a pull request, if that's any use? Is there a guide to contributing to Bolt anywhere?

```
commit 3a64d5210845af1e877e856e499920a7f7b0f198
Author: Mike McGowan <mike.mcgowan.75@gmail.com>
Date:   Thu May 11 14:57:37 2017 +0100

    Set default value of JSON columns to be {} rather than empty string (to keep PostgreSQL happy)

diff --git a/src/Storage/Entity/ContentValuesTrait.php b/src/Storage/Entity/ContentValuesTrait.php
index 2e39239..99bed95 100644
--- a/src/Storage/Entity/ContentValuesTrait.php
+++ b/src/Storage/Entity/ContentValuesTrait.php
@@ -150,7 +150,7 @@ trait ContentValuesTrait
                         if (!empty($this->values[$field]['url'])) {
                             $newvalue[$field] = json_encode($this->values[$field]);
                         } else {
-                            $newvalue[$field] = '';
+                            $newvalue[$field] = '{}';
                         }
                         break;
 
@@ -158,7 +158,7 @@ trait ContentValuesTrait
                         if (!empty($this->values[$field]['latitude']) && !empty($this->values[$field]['longitude'])) {
                             $newvalue[$field] = json_encode($this->values[$field]);
                         } else {
-                            $newvalue[$field] = '';
+                            $newvalue[$field] = '{}';
                         }
                         break;
 
@@ -166,7 +166,7 @@ trait ContentValuesTrait
                         if (!empty($this->values[$field]['file'])) {
                             $newvalue[$field] = json_encode($this->values[$field]);
                         } else {
-                            $newvalue[$field] = '';
+                            $newvalue[$field] = '{}';
                         }
                         break;
 
@@ -176,7 +176,7 @@ trait ContentValuesTrait
                             $newvalue[$field] = json_encode($this->values[$field]);
                         } elseif (!empty($this->values[$field]) && strlen($this->values[$field]) < 3) {
                             // Don't store '[]'
-                            $newvalue[$field] = '';
+                            $newvalue[$field] = '{}';
                         }
                         break;
 
@@ -209,7 +209,7 @@ trait ContentValuesTrait
             if (!empty($this['templatefields'])) {
                 $newvalue['templatefields'] = json_encode($this->values['templatefields']->getValues(true, true));
             } else {
-                $newvalue['templatefields'] = '';
+                $newvalue['templatefields'] = '{}';
             }
         }
```

(And it doesn't cover `selectmultiple` either.)

**Comment by GawainLynch** [comment]
So, I was wondering why I was seeing different results … I linked the wrong PR, in 3.4-alpha the prefill code was all rewritten in #6550 (moved to its own service, and new storage layer). The PR #6548 was the last Postgres fix that I hit … it works now, but I did just find an inconsistency I need to patch now

Now for < 3.4 …

Normal record editing through the backend UI won't use the code path, as it handles everything differently. 

The change above does not match how the new storage layer (plus interference from jQuery) saves empty fields. The JSON fields are nullable, and should be that, with the exception of `imagelist` `filelist`, & `templatefields`, where it is an empty JSON array, not object, i.e. `[]` 

> I could open a pull request, if that's any use? Is there a guide to contributing to Bolt anywhere?

We'd love you to PR this. there a page [Contributing to Bolt](https://docs.bolt.cm/other/contributing) on our docs site.

Just note my comments above and we should be good to go! :+1:
